File: python/kya/src/kya/_http.py
"""Internal HTTP transport layer for the KYA SDK."""

from __future__ import annotations

import time
from typing import Any

import httpx

from .errors import (
    KyaApiError,
    KyaAuthError,
    KyaNetworkError,
    KyaNotFoundError,
    KyaRateLimitError,
)

DEFAULT_BASE_URL = "https://pheme.ca/api/v1"
DEFAULT_TIMEOUT = 30.0
DEFAULT_MAX_RETRIES = 3


def _build_headers(
    api_key: str | None = None,
    jwt: str | None = None,
) -> dict[str, str]:
    headers: dict[str, str] = {
        "Accept": "application/json",
        "User-Agent": "kya-sdk-python/0.1.0",
    }
    if api_key:
        headers["X-API-Key"] = api_key
    if jwt:
        headers["Authorization"] = f"Bearer {jwt}"
    return headers


def _raise_for_status(response: httpx.Response) -> None:
    """Map HTTP error codes to typed SDK exceptions."""
    if response.is_success:
        return

    body: str | None = None
    try:
        body = response.text
    except Exception:
        pass

    if response.status_code == 401:
        raise KyaAuthError(body)

    if response.status_code == 404:
        raise KyaNotFoundError(body=body)

    if response.status_code == 429:
        retry_after = float(response.headers.get("Retry-After", "1"))
        raise KyaRateLimitError(retry_after=retry_after, body=body)

    try:
        detail = response.json().get("detail") or response.json().get("error") or response.text
    except Exception:
        detail = response.text or response.reason_phrase

    raise KyaApiError(response.status_code, str(detail), body)


# ─── Sync transport ──────────────────────────────────────────────────────────

class SyncTransport:
    def __init__(
        self,
        base_url: str = DEFAULT_BASE_URL,
        api_key: str | None = None,
        jwt: str | None = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._jwt = jwt
        self._timeout = timeout
        self._max_retries = max_retries
        self._client = httpx.Client(
            base_url=self._base_url,
            headers=_build_headers(api_key, jwt),
            timeout=timeout,
        )

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> SyncTransport:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

    def get(self, path: str, params: dict[str, Any] | None = None) -> Any:
        return self._request("GET", path, params=params)

    def _request(
        self,
        method: str,
        path: str,
        params: dict[str, Any] | None = None,
        json: Any | None = None,
    ) -> Any:
        attempts = 0
        last_exc: Exception | None = None

        while attempts < self._max_retries:
            try:
                response = self._client.request(method, path, params=params, json=json)
                _raise_for_status(response)
                return response.json()
            except KyaRateLimitError as exc:
                attempts += 1
                last_exc = exc
                if attempts < self._max_retries:
                    time.sleep(exc.retry_after)
            except (httpx.TimeoutException, httpx.ConnectError, httpx.NetworkError) as exc:
                raise KyaNetworkError(str(exc)) from exc
            except (KyaApiError,):
                raise

        assert last_exc is not None
        raise last_exc


# ─── Async transport ─────────────────────────────────────────────────────────

class AsyncTransport:
    def __init__(
        self,
        base_url: str = DEFAULT_BASE_URL,
        api_key: str | None = None,
        jwt: str | None = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._jwt = jwt
        self._timeout = timeout
        self._max_retries = max_retries
        self._client = httpx.AsyncClient(
            base_url=self._base_url,
            headers=_build_headers(api_key, jwt),
            timeout=timeout,
        )

    async def aclose(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> AsyncTransport:
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.aclose()

    async def get(self, path: str, params: dict[str, Any] | None = None) -> Any:
        return await self._request("GET", path, params=params)

    async def _request(
        self,
        method: str,
        path: str,
        params: dict[str, Any] | None = None,
        json: Any | None = None,
    ) -> Any:
        import asyncio

        attempts = 0
        last_exc: Exception | None = None

        while attempts < self._max_retries:
            try:
                response = await self._client.request(method, path, params=params, json=json)
                _raise_for_status(response)
                return response.json()
            except KyaRateLimitError as exc:
                attempts += 1
                last_exc = exc
                if attempts < self._max_retries:
                    await asyncio.sleep(exc.retry_after)
            except (httpx.TimeoutException, httpx.ConnectError, httpx.NetworkError) as exc:
                raise KyaNetworkError(str(exc)) from exc
            except (KyaApiError,):
                raise

        assert last_exc is not None
        raise last_exc

File: typescript/kya/src/http.ts
/**
 * KYA SDK — HTTP transport layer with retry, auth, and timeout support.
 */

import { KyaApiError, KyaAuthError, KyaNetworkError, KyaNotFoundError, KyaRateLimitError } from "./errors.js";

export interface RequestOptions {
  method?: string;
  body?: unknown;
  headers?: Record<string, string>;
  signal?: AbortSignal;
}

export interface HttpClientConfig {
  baseUrl: string;
  apiKey?: string;
  token?: string;
  timeout: number;
  maxRetries: number;
}

/**
 * Sleep for the given number of milliseconds.
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Parse the Retry-After header value into a millisecond delay.
 * Handles both delta-seconds and HTTP-date formats.
 */
function parseRetryAfter(header: string | null): number {
  if (!header) return 1000;
  const seconds = parseInt(header, 10);
  if (!isNaN(seconds)) return Math.max(seconds, 1) * 1000;
  // HTTP-date format
  const date = new Date(header);
  if (!isNaN(date.getTime())) {
    return Math.max(date.getTime() - Date.now(), 0);
  }
  return 1000;
}

/**
 * Minimal HTTP client for the KYA API.
 * Handles auth headers, JSON serialization, error mapping, and retry on 429.
 */
export class HttpClient {
  private readonly config: HttpClientConfig;

  constructor(config: HttpClientConfig) {
    this.config = config;
  }

  /**
   * Build auth headers based on configured credentials.
   * JWT token takes precedence over API key.
   */
  private authHeaders(): Record<string, string> {
    if (this.config.token) {
      return { Authorization: `Bearer ${this.config.token}` };
    }
    if (this.config.apiKey) {
      return { "X-API-Key": this.config.apiKey };
    }
    return {};
  }

  /**
   * Execute an HTTP request against the given path.
   * Automatically retries on 429 responses up to `maxRetries` times.
   */
  async request<T>(path: string, options: RequestOptions = {}, handle?: string): Promise<T> {
    const url = `${this.config.baseUrl}${path}`;
    const method = options.method ?? "GET";

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...this.authHeaders(),
      ...options.headers,
    };

    let attempt = 0;
    const maxAttempts = this.config.maxRetries + 1;

    for (let _i = 0; _i < maxAttempts + 1; _i++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

      let response: Response;

      try {
        response = await fetch(url, {
          method,
          headers,
          body: options.body !== undefined ? JSON.stringify(options.body) : undefined,
          signal: options.signal ?? controller.signal,
        });
      } catch (err) {
        clearTimeout(timeoutId);
        if (err instanceof Error && err.name === "AbortError") {
          throw new KyaNetworkError(`Request timed out after ${this.config.timeout}ms`, err);
        }
        throw new KyaNetworkError("Network error — could not reach the KYA service", err);
      } finally {
        clearTimeout(timeoutId);
      }

      // Handle rate limiting with retry
      if (response.status === 429) {
        const retryAfterHeader = response.headers.get("Retry-After");
        const retryAfterSeconds = parseInt(retryAfterHeader ?? "1", 10) || 1;
        const retryAfterMs = parseRetryAfter(retryAfterHeader);

        if (attempt < this.config.maxRetries) {
          attempt++;
          await sleep(retryAfterMs);
          continue;
        }

        const body = await response.text().catch(() => undefined);
        throw new KyaRateLimitError(retryAfterSeconds, body);
      }

      // Should never be reached, but satisfies TypeScript
      /* istanbul ignore next */
      if (_i >= maxAttempts) break;

      // Map HTTP error codes to typed errors
      if (!response.ok) {
        const body = await response.text().catch(() => undefined);

        switch (response.status) {
          case 401:
          case 403:
            throw new KyaAuthError(response.status as 401 | 403, body);
          case 404:
            throw new KyaNotFoundError(handle, body);
          default:
            throw new KyaApiError(response.status, response.statusText, body);
        }
      }

      // Parse JSON response
      try {
        return (await response.json()) as T;
      } catch (err) {
        throw new KyaNetworkError("Failed to parse JSON response", err);
      }
    }
  }

  /**
   * Fetch a raw text/SVG response (used for card SVG endpoint).
   */
  async requestText(path: string, options: RequestOptions = {}): Promise<string> {
    const url = `${this.config.baseUrl}${path}`;
    const method = options.method ?? "GET";

    const headers: Record<string, string> = {
      Accept: "image/svg+xml,text/plain,*/*",
      ...this.authHeaders(),
      ...options.headers,
    };

    let attempt = 0;
    const maxAttempts = this.config.maxRetries + 1;

    for (let _i = 0; _i < maxAttempts + 1; _i++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

      let response: Response;

      try {
        response = await fetch(url, {
          method,
          headers,
          signal: options.signal ?? controller.signal,
        });
      } catch (err) {
        clearTimeout(timeoutId);
        if (err instanceof Error && err.name === "AbortError") {
          throw new KyaNetworkError(`Request timed out after ${this.config.timeout}ms`, err);
        }
        throw new KyaNetworkError("Network error — could not reach the KYA service", err);
      } finally {
        clearTimeout(timeoutId);
      }

      if (response.status === 429) {
        const retryAfterHeader = response.headers.get("Retry-After");
        const retryAfterMs = parseRetryAfter(retryAfterHeader);
        if (attempt < this.config.maxRetries) {
          attempt++;
          await sleep(retryAfterMs);
          continue;
        }
        const retryAfterSeconds = parseInt(retryAfterHeader ?? "1", 10) || 1;
        throw new KyaRateLimitError(retryAfterSeconds);
      }

      /* istanbul ignore next */
      if (_i >= maxAttempts) break;

      if (!response.ok) {
        const body = await response.text().catch(() => undefined);
        switch (response.status) {
          case 401:
          case 403:
            throw new KyaAuthError(response.status as 401 | 403, body);
          case 404:
            throw new KyaNotFoundError(undefined, body);
          default:
            throw new KyaApiError(response.status, response.statusText, body);
        }
      }

      return response.text();
    }
  }
}
